services/wallet/activity/session.go

```diff
@@ -0,0 +1,313 @@
+		res := FilterResponse{
+			ErrorCode: ErrorCodeFailed,
+		}
+
+		if errors.Is(err, context.Canceled) || errors.Is(err, async.ErrTaskOverwritten) {
+			res.ErrorCode = ErrorCodeTaskCanceled
+		} else if err == nil {
+			activities := result.([]Entry)
+			res.Activities = activities
+			res.Offset = 0
+			res.HasMore = len(activities) == count
+			res.ErrorCode = ErrorCodeSuccess
+
+			processResults(activities)
+		}
+
+		int32SessionID := int32(f.sessionID)
+		sendResponseEvent(s.eventFeed, &int32SessionID, EventActivityFilteringDone, res, err)
+
+		s.getActivityDetailsAsync(int32SessionID, res.Activities)
+	})
+}
+
+func (s *Service) StartFilterSession(addresses []eth.Address, allAddresses bool, chainIDs []common.ChainID, filter Filter, firstPageCount int) SessionID {
+	sessionID := s.nextSessionID()
+
+	// TODO #12120: sort rest of the filters
+	// TODO #12120: prettyfy this
+	slices.SortFunc(addresses, func(a eth.Address, b eth.Address) bool {
+		return a.Hex() < b.Hex()
+	})
+	slices.Sort(chainIDs)
+	slices.SortFunc(filter.CounterpartyAddresses, func(a eth.Address, b eth.Address) bool {
+		return a.Hex() < b.Hex()
+	})
+
+	s.sessionsRWMutex.Lock()
+	subscribeToEvents := len(s.sessions) == 0
+	s.sessions[sessionID] = &Session{
+		id: sessionID,
+
+		addresses:    addresses,
+		allAddresses: allAddresses,
+		chainIDs:     chainIDs,
+		filter:       filter,
+
+		model: make([]EntryIdentity, 0, firstPageCount),
+	}
+	if subscribeToEvents {
+		s.subscribeToEvents()
+	}
+	s.sessionsRWMutex.Unlock()
+
+	s.internalFilter(fullFilter{
+		sessionID:    sessionID,
+		addresses:    addresses,
+		allAddresses: allAddresses,
+		chainIDs:     chainIDs,
+		filter:       filter,
+	}, 0, firstPageCount, func(entries []Entry) {
+		// Mirror identities for update use
+		s.sessionsRWMutex.Lock()
+		session, ok := s.sessions[sessionID]
+		if ok {
+			session.model = make([]EntryIdentity, 0, len(entries))
+			for _, a := range entries {
+				session.model = append(session.model, EntryIdentity{
+					payloadType: a.payloadType,
+					transaction: a.transaction,
+					id:          a.id,
+				})
+			}
+		}
+		s.sessionsRWMutex.Unlock()
+	})
+
+	return sessionID
+}
+
+// TODO #12120: extend the session based API
+//func (s *Service) GetMoreForFilterSession(count int) {}
+
+// subscribeToEvents should be called with sessionsRWMutex locked for writing
+func (s *Service) subscribeToEvents() {
+	s.ch = make(chan walletevent.Event)
+	s.subscriptions = s.eventFeed.Subscribe(s.ch)
+	go s.processEvents()
+}
+
+func (s *Service) processEvents() {
+	for event := range s.ch {
+		if event.Type == transactions.EventPendingTransactionUpdate {
+			var p transactions.PendingTxUpdatePayload
+			err := json.Unmarshal([]byte(event.Message), &p)
+			if err != nil {
+				log.Error("Error unmarshalling PendingTxUpdatePayload", "error", err)
+				continue
+			}
+
+			for id := range s.sessions {
+				s.sessionsRWMutex.RLock()
```

Comment:
This will work but "hide" the locking at lower function call levels. I try not to spread the mutexes at all call levels and keep it as high as possible. I plan to go away from locking in the future PRs.